src/doc/trpl/choosing-guarantees.md

```diff
@@ -0,0 +1,417 @@
+
+#### Guarantees
+
+Like `Rc`, this provides the (thread safe) guarantee that the destructor for the internal data will
+be run when the last `Arc` goes out of scope (barring any cycles).
+
+#### Cost
+
+This has the added cost of using atomics for changing the refcount (which will happen whenever it is
+cloned or goes out of scope). When sharing data from an `Arc` in a single thread, it is preferable
+to share `&` pointers whenever possible.
+
+[arc]: ../std/sync/struct.Arc.html
+
+## `Mutex<T>` and `RwLock<T>`
+
+[`Mutex<T>`][mutex] and [`RwLock<T>`][rwlock] provide mutual-exclusion via RAII guards. For both of
+these, the mutex is opaque until one calls `lock()` on it, at which point the thread will block
+until a lock can be acquired, and then a guard will be returned. This guard can be used to access
+the inner data (mutably), and the lock will be released when the guard goes out of scope.
+
+
+```rust
+{
+    let guard = mutex.lock();
+    // guard dereferences mutably to the inner type
+    *guard += 1;
+} // lock released when destructor runs
+```
+
+
+`RwLock` has the added benefit of being efficient for multiple reads. It is always safe to have
+multiple readers to shared data as long as there are no writers; and `RwLock` lets readers acquire a
+"read lock". Such locks can be acquired concurrently and are kept track of via a reference count.
+Writers must obtain a "write lock" which can only be obtained when all readers have gone out of
+scope.
+
+
+#### Guarantees
+
+Both of these provide safe shared mutability across threads, however they are prone to deadlocks.
+Some level of additional protocol safety can be obtained via the type system. An example  of this is
+[rust-sessions][sessions], an experimental library which uses session types for protocol safety.
+
+#### Costs
+
+These use internal atomic-like types to maintain the locks, and these are similar pretty costly
+(they can block all memory reads across processors till they're done). Waiting on these locks can
+also be slow when there's a lot of concurrent access happening.
+
+
+
+[rwlock]: ../std/sync/struct.RwLock.html
+[mutex]: ../std/sync/struct.Mutex.html
+[sessions]: https://github.com/Munksgaard/rust-sessions
+
+# Composition
+
+
+A common gripe when reading Rust code is with stuff like `Rc<RefCell<Vec<T>>>` and more complicated
+compositions of such types.
+
+Usually, it's a case of composing together the guarantees that one needs, without paying for stuff
+that is unnecessary.
+
+For example, `Rc<RefCell<T>>` is one such composition. `Rc` itself can't be dereferenced mutably;
+because `Rc` provides sharing and shared mutability isn't good, so we put `RefCell` inside to get
+dynamically verified shared mutability. Now we have shared mutable data, but it's shared in a way
```

Comment:
> because `Rc` provides sharing and shared mutability isn't good, so we put `RefCell` inside

Maybe "because `Rc` provides sharing (and shared mutability makes software complicated), so we put `RefCell` inside"